samples/order/processor/processor.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A processor processes images in orders. It receives requests by
+// subscribing to the requests topic, and writes responses to the response
+// topic.
+//
+// At present, it just encodes input images into PNG format.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"image"
+	_ "image/jpeg"
+	"image/png"
+	"log"
+	"time"
+
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	"gocloud.dev/docstore"
+	_ "gocloud.dev/docstore/memdocstore"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/mempubsub"
+	"gocloud.dev/samples/order/internal/common"
+)
+
+var (
+	requestSubURL    = flag.String("request-sub", "", "URL for request pub/sub subscription")
+	responseTopicURL = flag.String("response-topic", "", "URL for response pub/sub topic")
+	bucketURL        = flag.String("bucket", "", "URL for image bucket")
+	collectionURL    = flag.String("collection", "", "URL for order collection")
+)
+
+func main() {
+	flag.Parse()
+	if *requestSubURL == "" {
+		log.Fatal("missing -request-sub")
+	}
+	if *responseTopicURL == "" {
+		log.Fatal("missing -response-topic")
+	}
+	if *bucketURL == "" {
+		log.Fatal("missing -bucket")
+	}
+	if *collectionURL == "" {
+		log.Fatal("missing -collection")
+	}
+	ctx := context.Background()
+	p, err := newProcessor(ctx, *requestSubURL, *responseTopicURL, *bucketURL, *collectionURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := p.run(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type processor struct {
+	requestSub    *pubsub.Subscription
+	responseTopic *pubsub.Topic
+	bucket        *blob.Bucket
+	coll          *docstore.Collection
+}
+
+func newProcessor(ctx context.Context, subURL, topicURL, bucketURL, collURL string) (*processor, error) {
```

Comment:
Wouldn't it be better to do the URL -> portable type in `main`, and take the portable types here? Then `main` can `defer foo.Close()` etc. It's closer to real dependency injection as well.

And `newProcessor` would never fail.

Reply:
This is exists only to simplify the tests. (It's only called once from real code and once from tests.) I have no other architectural or code-quality goals.